### ISSUE TITLE
How to use multiple models for detection at the same time?

### ISSUE THREAD

**Issue by shizhanhao** [ultralytics/yolov5#4673]
Hello, Dear developers. Can you tell me how to merge models trained from multiple datasets together? Let's say I trained a model with a traffic light dataset, another model with a car dataset, and another model with a pedestrian dataset. So how do I reason with all three models at the same time to achieve recognition of traffic lights, cars, and pedestrians?

**Comment by kinoute** [comment]
Why not merge these models/datasets into one?

**Comment by shizhanhao** [comment]
> Why not merge these models/datasets into one?
First of all, thank you for your reply, because each dataset has tens of thousands of photos, and there is only one category of annotation in each dataset. For example, most of the photos in the vehicle dataset have traffic lights on them, but they are not labeled, so if multiple data sets are directly combined together, it will affect the map value, and the generalization of the model trained in this way is not good.

**Comment by kinoute** [comment]
Could you not use your traffic lights model to generate annotations files for the vehicle dataset? (to annotate the traffic lights)

That's what I did. I trained a small model made quickly to pseudo-annotate the rest of my dataset. I put my model as an API service, got the predictions/normalized coordinates, and generated the annotation text files from the result, for every image.

After that, you just have to double-check if the predicted coordinates are correct (to remove some false positive/negative), but it's quick. I checked 10k images in around 2h.

**Comment by shizhanhao** [comment]
> Could you not use your traffic lights model to generate annotations files for the vehicle dataset? (to annotate the traffic lights)
> 
> That's what I did. I trained a small model made quickly to pseudo-annotate the rest of my dataset. I put my model as an API service, got the predictions/normalized coordinates, and generated the annotation text files from the result, for every image.
> 
> After that, you just have to double-check if the predicted coordinates are correct (to remove some false positive/negative), but it's quick. I checked 10k images in around 2h.



> Could you not use your traffic lights model to generate annotations files for the vehicle dataset? (to annotate the traffic lights)
> 
> That's what I did. I trained a small model made quickly to pseudo-annotate the rest of my dataset. I put my model as an API service, got the predictions/normalized coordinates, and generated the annotation text files from the result, for every image.
> 
> After that, you just have to double-check if the predicted coordinates are correct (to remove some false positive/negative), but it's quick. I checked 10k images in around 2h.



> Could you not use your traffic lights model to generate annotations files for the vehicle dataset? (to annotate the traffic lights)
> 
> That's what I did. I trained a small model made quickly to pseudo-annotate the rest of my dataset. I put my model as an API service, got the predictions/normalized coordinates, and generated the annotation text files from the result, for every image.
> 
> After that, you just have to double-check if the predicted coordinates are correct (to remove some false positive/negative), but it's quick. I checked 10k images in around 2h.

This method can indeed be used, but the workload is still a bit large, thank you for your reply and support.

**Comment by sekisek** [comment]
I have the same questions but in a different scenario.
Is it possible to load multiple models to the same page (detect.py) ???

**Comment by glenn-jocher** [comment]
@sekisek yes you can run inference with multiple models at the same time. See **Model Ensembling** Tutorial below:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by sekisek** [comment]
Hi @glenn-jocher ,
thanks, I read it!
but can you combine 2 different models  for example like car-plate-detection with 1 class and the regular yolov5x.pt
or between any 2 customer models with different classes?

**Comment by shizhanhao** [comment]
I have tried, the two models must be a data set to use the method described by the author, which means that two models from different data sets cannot be used together, but you can use different models for the same photo Perform multiple inferences.

**Comment by minhN2000** [comment]
hi @shizhanhao,
may I ask what commands you need to add in order to have a photo uses different models? 
Thank you!

**Comment by vayvay0993** [comment]
@shizhanhao I have the exactly same question with you, just want to know if you have the solution for this issue?

**Comment by jjjonathan96** [comment]
My question is i have trained YOLOv5 model for hand held object detection and there are some model for face detection in YOLO so, I need to have a single model detecting faces and hand held objects. Is it possible to have singular model? Will meta-learning make sense?

**Comment by glenn-jocher** [comment]
See https://community.ultralytics.com/t/how-to-combine-weights-to-detect-from-multiple-datasets

**Comment by hemalbeselial** [comment]
hi i have a question i using a yolov5n model for object detection and i have a another code for face recognition now i add the face recognition code to detect.py code. when i run my code give me 2 separate result windows but i need both the face recognition and yolo object detection should happen in the same screen how can i do it

**Comment by glenn-jocher** [comment]
@hemalbeselial you can use the `detect.py` script to run both the face recognition and YOLO object detection in the same screen. You can simply integrate both models into the same script and define the necessary logic to display the results together in a single output screen.
